### ISSUE TITLE
autoCommit: false does not work as expected with Oracle Client v 11.2

### ISSUE THREAD

**Issue by fedulovivan** [oracle/node-oracledb#269]
Hello! Please consider following scenario:

We have:
- oracledb v 1.3 compiled with instant client v 11.2.0.4
- oracle dbms 11.2.0.2.0 Enterprise Edition
- connection pool is used
- autoCommit option set via execute call

I make attempt to run two INSERT statements using single connection with option *autoCommit* set to *false* for both of them (passed via execute). Second query is intentionally erroneous (have a foreign key value violating constraint). On catching error for second query I release connection back to pool. No manual rollback performed. Checking db tables shows that first query actually affected database, row is in place, but supposed to be rolled back automatically.

Note that this works just fine when module is compiled with instant client v 12.1.0.2

**Comment by cjbj** [comment]
- What platform(s) are you on?
- Please share your testcase so we can make sure we are testing what you are testing.

**Comment by fedulovivan** [comment]
This was tested on 64 bit Ubuntu 14 LTS.
Please find test script below:
```javascript
var oracledb = require("oracledb");

console.log("oracleClientVersion=", oracledb.oracleClientVersion);

oracledb.createPool({
    "user": "****",
    "password": "*****",
    "connectString": "****/****"
}, function(err, pool) {
    if(err) console.log(err);
    pool.getConnection(function(err, conn) {
        if(err) console.log(err);
        console.log("oracleServerVersion=", conn.oracleServerVersion);
        // sql will be the same for both execute calls
        var procSql = "BEGIN pkg_******.add_******(p_iname=>:p_iname, p_short_name=>:p_short_name, p_comments=>:p_comments, p_new_id=>:p_new_id, p_status=>:p_status, p_description=>:p_description); END;";
        // execute queries one by one
        // first one
        conn.execute(
            procSql,
            {
                p_iname: "Test iname",
                p_short_name: "TST",
                p_comments: "Test comments",
                p_new_id: {
                    type: oracledb.NUMBER,
                    dir: oracledb.BIND_OUT
                },
                p_status: {
                    type: oracledb.NUMBER,
                    dir: oracledb.BIND_OUT
                },
                p_description: {
                    type: oracledb.STRING,
                    dir: oracledb.BIND_OUT
                }
            },
            {
                autoCommit: false,
                resultSet: false
            },
            function(err, results) {
                if(err) {
                    console.log(err);
                    return;
                }
                // second one
                conn.execute(
                    procSql,
                    {
                        p_iname123: "Test iname", // specify wrong bind parameter name to cause an error
                        p_short_name: "TST",
                        p_comments: "Test comments",
                        p_new_id: {
                            type: oracledb.NUMBER,
                            dir: oracledb.BIND_OUT
                        },
                        p_status: {
                            type: oracledb.NUMBER,
                            dir: oracledb.BIND_OUT
                        },
                        p_description: {
                            type: oracledb.STRING,
                            dir: oracledb.BIND_OUT
                        }
                    },
                    {
                        autoCommit: false,
                        resultSet: false
                    },
                    function(err, results) {
                        if(err) {
                            if(err) console.log(err);
                            conn.release(function(err) {
                                if(err) console.log(err); else console.log("released");
                            });
                        } else {
                            conn.commit(function(err) {
                                if(err) console.log(err); else console.log("comitted");
                            });
                        }
                    }
                );
            }
        );
    });
});
```
It will produce following output:
```bash
**** ~/Desktop/Projects/***** $ node issue-269.js
oracleClientVersion= 1102000400
oracleServerVersion= 1102000200
[Error: ORA-01036: illegal variable name/number]
released
```
Corresponding db table check shows that first insert was successful and row is in place.

**Comment by dmcghan** [comment]
@fedulovivan Can your provide pkg_******.add_******? I'm guessing you can't because you obfuscated the name. If not, can you test with just insert statements instead of a procedure call?

Also, can you double verify that your proc isn't doing a commit? Is it executing any DDL? That would involve implicit commits...

**Comment by fedulovivan** [comment]
@dmcghan Unfortunately I cannot provide source code of this procedure. In fact, as I already mentioned, auto rollback works fine when oracledb module is compiled with newer version of instant client library (12.1.0.2). I don't think the issue is connected with procedure code itself. Will try the same with plain insert queries and update with results.

**Comment by cjbj** [comment]
I can reproduce with
```javascript
/*

drop table issue269tab;
create table issue269tab (myts timestamp, p_iname VARCHAR2(40), p_short_name VARCHAR2(40), p_comments VARCHAR2(40));
create or replace procedure issue269proc(p_iname IN VARCHAR2, p_short_name IN VARCHAR2, p_comments IN VARCHAR2, p_new_id OUT NUMBER, p_status OUT NUMBER, p_description OUT VARCHAR2)
as
BEGIN
p_description := p_iname || ' ' || p_short_name || ' ' || p_comments;
p_new_id := 1;
p_status := 2;
insert into issue269tab values (systimestamp, p_iname, p_short_name, p_comments);
END;
/
show errors

 */

'use strict';

var oracledb = require("oracledb");
var dbconfig = require('./dbconfig.js');

//oracledb.autoCommit = false;

console.log("oracleClientVersion=", oracledb.oracleClientVersion);

oracledb.createPool(
  dbconfig,
  function(err, pool) {
    if(err) console.log(err);
    pool.getConnection(function(err, conn) {
      if(err) { console.log(err); return; }
      console.log("oracleServerVersion=", conn.oracleServerVersion);
      // sql will be the same for both execute calls
      var procSql = "BEGIN issue269proc(p_iname=>:p_iname, p_short_name=>:p_short_name, p_comments=>:p_comments, p_new_id=>:p_new_id, p_status=>:p_status, p_description=>:p_description); END;";
      // execute queries one by one
      // first one
      conn.execute(
        procSql,
        {
          p_iname: "Test iname",
          p_short_name: "TST",
          p_comments: "Test comments",
          p_new_id: {
            type: oracledb.NUMBER,
            dir: oracledb.BIND_OUT
          },
          p_status: {
            type: oracledb.NUMBER,
            dir: oracledb.BIND_OUT
          },
          p_description: {
            type: oracledb.STRING,
            dir: oracledb.BIND_OUT
          }
        },
        {
          autoCommit: false,
          resultSet: false
        },
        function(err, results) {
          if(err) {
            console.log(err);
            return;
          }
          // second one
          conn.execute(
            procSql,
            {
              p_iname123: "Test iname", // specify wrong bind parameter name to cause an error
              p_short_name: "TST",
              p_comments: "Test comments",
              p_new_id: {
                type: oracledb.NUMBER,
                dir: oracledb.BIND_OUT
              },
              p_status: {
                type: oracledb.NUMBER,
                dir: oracledb.BIND_OUT
              },
              p_description: {
                type: oracledb.STRING,
                dir: oracledb.BIND_OUT
              }
            },
            {
              autoCommit: false,
              resultSet: false
            },
            function(err, results) {
              if(err) {
                if(err) console.log(err);
                conn.release(function(err) {
                  if(err) console.log(err); else console.log("released");
                });
              } else {
                conn.commit(function(err) {
                  if(err) console.log(err); else console.log("comitted");
                });
              }
            }
          );
        }
      );
    });
  });
```

We'll look into it.

**Comment by cjbj** [comment]
The natural behavior of OCI is to commit when a connection is released. This is the opposite of node-oracledb, which therefore has to determine whether to rollback or not. 

The reason the bug isn't present when node-oracledb is linked with 12c client libraries is [this code](https://github.com/oracle/node-oracledb/blob/2745240a0ac41b93851958f3907e4f590787eaa4/src/dpi/src/dpiConnImpl.cpp#L188).

When node-oracledb is linked with 11g client we have to use a [heuristic](https://github.com/oracle/node-oracledb/blob/2745240a0ac41b93851958f3907e4f590787eaa4/src/dpi/src/dpiStmtImpl.cpp#L351). This  needs to be changed, which will cause more, sometimes unnecessary, rollbacks to be issued.  

Bottom line: use 12c client libraries if possible.

**Comment by fedulovivan** [comment]
@cjbj Thanks for detailed explanation! In fact so far we ought to use namely 12.1 due to this issue. However, our company dba's recommendation was to use client library version which match server version (11.2 in our case). So, anyways, looking forward for this issue to be fixed some day.

**Comment by cjbj** [comment]
Fixed in 1.5.0
